### PR TITLE
TW-428 can invite internal matrix account to group chat in editing group

### PR DESCRIPTION
**Pull request by drminh2807**
# Issue
- #428 
- #431 
# Demo
## Invite search
https://github.com/linagora/twake-on-matrix/assets/12546908/945e3132-bd64-4bdb-9714-aca516ca7baa

## Invite external contact

https://github.com/linagora/twake-on-matrix/assets/12546908/30a3d8f2-1f29-4f84-9ce1-7adcde2c65ab

### CODE REVIEW

**Comment by hoangdat** [comment]
- [ ] show check box for external account also

File: lib/pages/new_group/contacts_selection.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+class ContactsSelectionController<T extends StatefulWidget> extends State<T>
+    with SearchContactsController, InviteExternalContactMixin {
+  final selectedContactsMapNotifier = SelectedContactsMapChangeNotifier();
+
+  String getTitle(BuildContext context) => "Should override";
+
+  String getHintText(BuildContext context) => "Should override";
+
+  void onSubmit() {}
```
Comment (sherlockvn) [inline lib/pages/new_group/contacts_selection.dart:17]: Why not use abstract

File: lib/pages/new_private_chat/search_contacts_controller.dart [hunk 2]
```diff
@@ -7,6 +7,7 @@ import 'package:fluffychat/pages/search/get_contacts_controller.dart';
 import 'package:fluffychat/presentation/converters/presentation_contact_converter.dart';
 import 'package:flutter/material.dart';
 import 'package:matrix/matrix.dart';
+import 'package:pull_to_refresh/pull_to_refresh.dart';
 
 mixin class SearchContactsController {
```
Comment (sherlockvn) [inline lib/pages/new_private_chat/search_contacts_controller.dart:12]: should have name end with mixin

File: lib/pages/invitation_selection/invitation_selection.dart [hunk 3]
```diff
@@ -20,119 +18,74 @@ class InvitationSelection extends StatefulWidget {
...
-  late String currentSearchTerm;
-  bool loading = false;
-  List<Profile> foundProfiles = [];
-  Timer? coolDown;
+class InvitationSelectionController
+    extends ContactsSelectionController<InvitationSelection> {
+  String? get _roomId => GoRouterState.of(context).pathParameters['roomid'];
 
-  String? get roomId => GoRouterState.of(context).pathParameters['roomid'];
+  Room get _room => Matrix.of(context).client.getRoomById(_roomId!)!;
```
Comment (hoangdat) [inline lib/pages/invitation_selection/invitation_selection.dart:25]: 👍

File: lib/pages/new_group/contacts_selection.dart (hunk 1 above)
Comment (drminh2807) [reply lib/pages/new_group/contacts_selection.dart:17]: great idea, fixed

File: lib/pages/new_group/contacts_selection.dart [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:fluffychat/presentation/model/presentation_contact.dart';
+import 'package:flutter/cupertino.dart';
+
+class ContactsSelectionController<T extends StatefulWidget> extends State<T>
+    with SearchContactsController, InviteExternalContactMixin {
+  final selectedContactsMapNotifier = SelectedContactsMapChangeNotifier();
+
+  String getTitle(BuildContext context) => "Should override";
+
+  String getHintText(BuildContext context) => "Should override";
```
Comment (hoangdat) [inline lib/pages/new_group/contacts_selection.dart:15]: @@, should be the abstract class :))

File: lib/pages/new_private_chat/search_contacts_controller.dart (hunk 2 above)
Comment (drminh2807) [reply lib/pages/new_private_chat/search_contacts_controller.dart:12]: it can still be used as a standalone controller

**Review by hoangdat — APPROVED** [review]

File: lib/pages/new_private_chat/search_contacts_controller.dart (hunk 2 above)
Comment (sherlockvn) [reply lib/pages/new_private_chat/search_contacts_controller.dart:12]: `SearchContactsMixinController`

File: lib/domain/model/contact/contact_status.dart [hunk 5]
```diff
@@ -1 +1 @@
-enum ContactStatus { inactive, active }
+enum ContactStatus { inactive, active, external }
```
Comment (sherlockvn) [inline lib/domain/model/contact/contact_status.dart:1]: external may be  inactive or active also ?

File: lib/mixin/invite_external_contact_mixin.dart [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+mixin InviteExternalContactMixin on SearchContactsController {
+  void initSearchExternalContacts() {
+    initSearchContacts(
+      converter: PresentationContactConverter(checkExternal: true),
+    );
+  }
+
+  void showInviteExternalContactDialog(
+    BuildContext context,
+    Function() onAccept,
```
Comment (sherlockvn) [inline lib/mixin/invite_external_contact_mixin.dart:16]: `void Function() onAccept`

File: lib/pages/invitation_selection/invitation_selection.dart (hunk 3 above)
Comment (sherlockvn) [reply lib/pages/invitation_selection/invitation_selection.dart:25]: should check null, right ?

File: lib/pages/new_group/contacts_selection_view.dart [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          ),
+          Expanded(
+            child: ValueListenableBuilder<bool>(
+              valueListenable: controller
+                  .selectedContactsMapNotifier.haveSelectedContactsNotifier,
+              builder: (context, haveSelectedContact, child) {
+                return Padding(
+                  padding: EdgeInsets.only(
+                    top: haveSelectedContact ? 0.0 : 8.0,
+                  ),
```
Comment (nqhhdev) [inline lib/pages/new_group/contacts_selection_view.dart:42]: Move it to `SearchableAppBarStyle`

File: lib/pages/new_group/new_group.dart [hunk 8]
```diff
@@ -59,33 +51,43 @@ class NewGroupController extends State<NewGroup>
...
+    return L10n.of(context)!.newGroupChat;
+  }
+
+  @override
+  String getHintText(BuildContext context) {
+    return L10n.of(context)!.whoWouldYouLikeToAdd;
+  }
+
+  @override
+  List<String> get disabledContacts => [];
```
Comment (sherlockvn) [inline lib/pages/new_group/new_group.dart:79]: What is disabledContacts, and you shoudn't override like this if it don't have meaning in this class, maybe don't need to abstract this method,

File: lib/widgets/twake_components/twake_smart_refresher.dart [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    required this.controller,
+    required this.onRefresh,
+    required this.onLoading,
+    this.child,
+  }) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return SmartRefresher(
+      enablePullUp: true,
```
Comment (nqhhdev) [inline lib/widgets/twake_components/twake_smart_refresher.dart:21]: IMO: Not set default is `true`. I think it's optional

File: lib/domain/model/contact/contact_status.dart (hunk 5 above)
Comment (hoangdat) [reply lib/domain/model/contact/contact_status.dart:1]: ohh, nice catches. Please use `ContactType`

File: lib/pages/new_private_chat/search_contacts_controller.dart (hunk 2 above)
Comment (drminh2807) [reply lib/pages/new_private_chat/search_contacts_controller.dart]: done

File: lib/pages/new_group/contacts_selection_view.dart (hunk 7 above)
Comment (drminh2807) [reply lib/pages/new_group/contacts_selection_view.dart]: done

File: lib/domain/model/contact/contact_status.dart (hunk 5 above)
Comment (drminh2807) [reply lib/domain/model/contact/contact_status.dart]: done

File: lib/mixin/invite_external_contact_mixin.dart (hunk 6 above)
Comment (drminh2807) [reply lib/mixin/invite_external_contact_mixin.dart]: done

File: lib/widgets/twake_components/twake_smart_refresher.dart (hunk 9 above)
Comment (drminh2807) [reply lib/widgets/twake_components/twake_smart_refresher.dart]: done

**Review by hoangdat — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**